File: tests/test_function.cpp
/**
 * Test: test_function
 * Tests LLVM Function creation and properties
 *
 * LLVM-C APIs covered:
 * - LLVMAddFunction()
 * - LLVMGetNamedFunction()
 * - LLVMCountParams(), LLVMGetParams(), LLVMGetParam()
 * - LLVMSetValueName2(), LLVMGetValueName2()
 * - LLVMGetFunctionCallConv(), LLVMSetFunctionCallConv()
 * - LLVMGetLinkage(), LLVMSetLinkage()
 * - LLVMGetFirstFunction(), LLVMGetNextFunction(), LLVMGetLastFunction()
 * - LLVMDeleteFunction()
 * - LLVMGetReturnType(), LLVMCountParamTypes()
 */

#include <cstdio>
#include <cstring>
#include <llvm-c/Analysis.h>
#include <llvm-c/Core.h>

const char *linkage_name(LLVMLinkage linkage) {
  switch (linkage) {
  case LLVMExternalLinkage:
    return "external";
  case LLVMAvailableExternallyLinkage:
    return "available_externally";
  case LLVMLinkOnceAnyLinkage:
    return "linkonce";
  case LLVMLinkOnceODRLinkage:
    return "linkonce_odr";
  case LLVMWeakAnyLinkage:
    return "weak";
  case LLVMWeakODRLinkage:
    return "weak_odr";
  case LLVMAppendingLinkage:
    return "appending";
  case LLVMInternalLinkage:
    return "internal";
  case LLVMPrivateLinkage:
    return "private";
  case LLVMExternalWeakLinkage:
    return "extern_weak";
  case LLVMCommonLinkage:
    return "common";
  default:
    return "unknown";
  }
}

int main() {
  LLVMContextRef ctx = LLVMContextCreate();
  LLVMModuleRef mod = LLVMModuleCreateWithNameInContext("test_function", ctx);

  LLVMTypeRef i32 = LLVMInt32TypeInContext(ctx);
  LLVMTypeRef i64 = LLVMInt64TypeInContext(ctx);
  LLVMTypeRef void_ty = LLVMVoidTypeInContext(ctx);
  LLVMTypeRef ptr = LLVMPointerTypeInContext(ctx, 0);

  // Function 1: void foo()
  LLVMTypeRef foo_ty = LLVMFunctionType(void_ty, nullptr, 0, 0);
  LLVMValueRef foo = LLVMAddFunction(mod, "foo", foo_ty);

  // Function 2: i32 bar(i32, i32)
  LLVMTypeRef bar_params[] = {i32, i32};
  LLVMTypeRef bar_ty = LLVMFunctionType(i32, bar_params, 2, 0);
  LLVMValueRef bar = LLVMAddFunction(mod, "bar", bar_ty);

  // Set parameter names
  LLVMValueRef bar_param0 = LLVMGetParam(bar, 0);
  LLVMValueRef bar_param1 = LLVMGetParam(bar, 1);
  LLVMSetValueName2(bar_param0, "x", 1);
  LLVMSetValueName2(bar_param1, "y", 1);

  // Function 3: i64 baz(ptr, i32, i64) with internal linkage
  // Internal linkage requires a body, so we add a simple one
  LLVMTypeRef baz_params[] = {ptr, i32, i64};
  LLVMTypeRef baz_ty = LLVMFunctionType(i64, baz_params, 3, 0);
  LLVMValueRef baz = LLVMAddFunction(mod, "baz", baz_ty);
  LLVMSetLinkage(baz, LLVMInternalLinkage);

  // Add a basic block with return to make it a valid definition
  LLVMBasicBlockRef baz_entry =
      LLVMAppendBasicBlockInContext(ctx, baz, "entry");
  LLVMBuilderRef builder = LLVMCreateBuilderInContext(ctx);
  LLVMPositionBuilderAtEnd(builder, baz_entry);
  LLVMBuildRet(builder, LLVMConstInt(i64, 0, 0));
  LLVMDisposeBuilder(builder);

  // Function 4: varargs function - i32 printf(ptr, ...)
  LLVMTypeRef printf_params[] = {ptr};
  LLVMTypeRef printf_ty = LLVMFunctionType(i32, printf_params, 1, 1);
  LLVMValueRef printf_fn = LLVMAddFunction(mod, "printf", printf_ty);

  // Function 5: Function with fastcc calling convention
  LLVMTypeRef fastcc_params[] = {i32};
  LLVMTypeRef fastcc_ty = LLVMFunctionType(i32, fastcc_params, 1, 0);
  LLVMValueRef fastcc_fn = LLVMAddFunction(mod, "fastcc_func", fastcc_ty);
  LLVMSetFunctionCallConv(fastcc_fn, LLVMFastCallConv);

  // Function 6: Will be deleted
  LLVMTypeRef delete_ty = LLVMFunctionType(void_ty, nullptr, 0, 0);
  LLVMValueRef delete_fn = LLVMAddFunction(mod, "to_be_deleted", delete_ty);

  // Get function by name
  LLVMValueRef found_bar = LLVMGetNamedFunction(mod, "bar");

  // Count functions before deletion
  int count_before = 0;
  for (LLVMValueRef fn = LLVMGetFirstFunction(mod); fn;
       fn = LLVMGetNextFunction(fn)) {
    count_before++;
  }

  // Delete the function
  LLVMDeleteFunction(delete_fn);

  // Count functions after deletion
  int count_after = 0;
  for (LLVMValueRef fn = LLVMGetFirstFunction(mod); fn;
       fn = LLVMGetNextFunction(fn)) {
    count_after++;
  }

  // Verify module
  char *error = nullptr;
  if (LLVMVerifyModule(mod, LLVMReturnStatusAction, &error)) {
    fprintf(stderr, "; Verification failed: %s\n", error);
    LLVMDisposeMessage(error);
    LLVMDisposeModule(mod);
    LLVMContextDispose(ctx);
    return 1;
  }
  LLVMDisposeMessage(error);

  // Print diagnostic comments
  printf("; Test: test_function\n");
  printf(";\n");

  // foo info
  size_t name_len;
  const char *foo_name = LLVMGetValueName2(foo, &name_len);
  printf("; Function 'foo':\n");
  printf(";   name: %s\n", foo_name);
  printf(";   param count: %u\n", LLVMCountParams(foo));
  printf(";   linkage: %s\n", linkage_name(LLVMGetLinkage(foo)));
  printf(";   calling conv: %u (C=0)\n", LLVMGetFunctionCallConv(foo));

  // bar info
  const char *bar_name = LLVMGetValueName2(bar, &name_len);
  printf(";\n");
  printf("; Function 'bar':\n");
  printf(";   name: %s\n", bar_name);
  printf(";   param count: %u\n", LLVMCountParams(bar));
  printf(";   found by name: %s\n", found_bar == bar ? "yes" : "no");

  // Get param names
  const char *p0_name = LLVMGetValueName2(bar_param0, &name_len);
  const char *p1_name = LLVMGetValueName2(bar_param1, &name_len);
  printf(";   param 0 name: %s\n", p0_name);
  printf(";   param 1 name: %s\n", p1_name);

  // baz info
  printf(";\n");
  printf("; Function 'baz':\n");
  printf(";   param count: %u\n", LLVMCountParams(baz));
  printf(";   linkage: %s\n", linkage_name(LLVMGetLinkage(baz)));

  // printf info
  printf(";\n");
  printf("; Function 'printf':\n");
  printf(";   param count: %u\n", LLVMCountParams(printf_fn));
  printf(";   is vararg: %s\n", LLVMIsFunctionVarArg(printf_ty) ? "yes" : "no");

  // fastcc info
  printf(";\n");
  printf("; Function 'fastcc_func':\n");
  printf(";   calling conv: %u (FastCall=8)\n",
         LLVMGetFunctionCallConv(fastcc_fn));

  // Function counts
  printf(";\n");
  printf("; Function count before deletion: %d\n", count_before);
  printf("; Function count after deletion: %d\n", count_after);

  // List all functions
  printf(";\n");
  printf("; All functions:\n");
  for (LLVMValueRef fn = LLVMGetFirstFunction(mod); fn;
       fn = LLVMGetNextFunction(fn)) {
    const char *fn_name = LLVMGetValueName2(fn, &name_len);
    printf(";   - %s\n", fn_name);
  }

  printf("\n");

  // Print module IR
  char *ir = LLVMPrintModuleToString(mod);
  printf("%s", ir);
  LLVMDisposeMessage(ir);

  // Cleanup
  LLVMDisposeModule(mod);
  LLVMContextDispose(ctx);

  return 0;
}

File: tests/test_function.py
#!/usr/bin/env -S uv run
"""
Test: test_function
Tests LLVM Function creation and properties via Python bindings.

This is the Python equivalent of tests/test_function.cpp.
Output should match the C++ golden master test.

LLVM APIs covered (via Python bindings):
- add_function(), get_function()
- param_count, get_param(), params
- name property (get/set)
- calling_conv property (get/set)
- linkage property (get/set)
- functions (module iteration)
- erase()
"""

import sys
import llvm


def linkage_name(linkage: llvm.Linkage) -> str:
    """Convert Linkage enum to the string name used in C++ test."""
    mapping = {
        llvm.Linkage.External: "external",
        llvm.Linkage.AvailableExternally: "available_externally",
        llvm.Linkage.LinkOnceAny: "linkonce",
        llvm.Linkage.LinkOnceODR: "linkonce_odr",
        llvm.Linkage.WeakAny: "weak",
        llvm.Linkage.WeakODR: "weak_odr",
        llvm.Linkage.Appending: "appending",
        llvm.Linkage.Internal: "internal",
        llvm.Linkage.Private: "private",
        llvm.Linkage.ExternalWeak: "extern_weak",
        llvm.Linkage.Common: "common",
    }
    return mapping.get(linkage, "unknown")


def main():
    with llvm.create_context() as ctx:
        with ctx.create_module("test_function") as mod:
            i32 = ctx.types.i32
            i64 = ctx.types.i64
            void_ty = ctx.types.void
            ptr = ctx.types.ptr

            # Function 1: void foo()
            foo_ty = ctx.types.function(void_ty, [], vararg=False)
            foo = mod.add_function("foo", foo_ty)

            # Function 2: i32 bar(i32, i32)
            bar_ty = ctx.types.function(i32, [i32, i32], vararg=False)
            bar = mod.add_function("bar", bar_ty)

            # Set parameter names
            bar_param0 = bar.get_param(0)
            bar_param1 = bar.get_param(1)
            bar_param0.name = "x"
            bar_param1.name = "y"

            # Function 3: i64 baz(ptr, i32, i64) with internal linkage
            # Internal linkage requires a body, so we add a simple one
            baz_ty = ctx.types.function(i64, [ptr, i32, i64], vararg=False)
            baz = mod.add_function("baz", baz_ty)
            baz.linkage = llvm.Linkage.Internal

            # Add a basic block with return to make it a valid definition
            baz_entry = baz.append_basic_block("entry")
            with baz_entry.create_builder() as builder:
                builder.ret(i64.constant(0))

            # Function 4: varargs function - i32 printf(ptr, ...)
            printf_ty = ctx.types.function(i32, [ptr], vararg=True)
            printf_fn = mod.add_function("printf", printf_ty)

            # Function 5: Function with fastcc calling convention
            fastcc_ty = ctx.types.function(i32, [i32], vararg=False)
            fastcc_fn = mod.add_function("fastcc_func", fastcc_ty)
            fastcc_fn.calling_conv = llvm.CallConv.Fast

            # Function 6: Will be deleted
            delete_ty = ctx.types.function(void_ty, [], vararg=False)
            delete_fn = mod.add_function("to_be_deleted", delete_ty)

            # Get function by name
            found_bar = mod.get_function("bar")

            # Count functions before deletion
            count_before = len(mod.functions)

            # Delete the function
            delete_fn.erase()

            # Count functions after deletion
            count_after = len(mod.functions)

            # Verify module
            if not mod.verify():
                print(
                    f"; Verification failed: {mod.get_verification_error()}",
                    file=sys.stderr,
                )
                return 1

            # Print diagnostic comments
            print("; Test: test_function")
            print(";")

            # foo info
            print("; Function 'foo':")
            print(f";   name: {foo.name}")
            print(f";   param count: {foo.param_count}")
            print(f";   linkage: {linkage_name(foo.linkage)}")
            print(f";   calling conv: {foo.calling_conv.value} (C=0)")

            # bar info
            print(";")
            print("; Function 'bar':")
            print(f";   name: {bar.name}")
            print(f";   param count: {bar.param_count}")
            print(f";   found by name: {'yes' if found_bar is not None else 'no'}")

            # Get param names
            print(f";   param 0 name: {bar_param0.name}")
            print(f";   param 1 name: {bar_param1.name}")

            # baz info
            print(";")
            print("; Function 'baz':")
            print(f";   param count: {baz.param_count}")
            print(f";   linkage: {linkage_name(baz.linkage)}")

            # printf info
            print(";")
            print("; Function 'printf':")
            print(f";   param count: {printf_fn.param_count}")
            print(f";   is vararg: {'yes' if printf_ty.is_vararg else 'no'}")

            # fastcc info
            print(";")
            print("; Function 'fastcc_func':")
            print(f";   calling conv: {fastcc_fn.calling_conv.value} (FastCall=8)")

            # Function counts
            print(";")
            print(f"; Function count before deletion: {count_before}")
            print(f"; Function count after deletion: {count_after}")

            # List all functions
            print(";")
            print("; All functions:")
            for fn in mod.functions:
                print(f";   - {fn.name}")

            print()

            # Print module IR
            print(mod.to_string(), end="")

    return 0


if __name__ == "__main__":
    sys.exit(main())
